exe12_Q2.Rmd

```diff
@@ -0,0 +1,66 @@
+
+```{r}
+library(stringr)
+
+all_hours <- c("07:30", "10:15", "10:90", "12:01","15:30","15:80", "20:01","23:59","24:00","25:01", "23:591", "123:45", "18:00")
+
+all_hours_answers <- vector(length = length(all_hours))
+
+for(i in 1:length(all_hours)){
+  if(str_detect(all_hours[i],"^1")==TRUE){ #1
+    if(str_detect(all_hours[i],"^12:0")){ #2
+      all_hours_answers[i] <- str_detect(all_hours[i],"^12:0[1-9]$") #3
+    }
+    else{
+      all_hours_answers[i] <- str_detect(all_hours[i],"^1[3-9]:[0-5][0-9]$") #4
+    }
+  }
+  else{
+    all_hours_answers[i] <- str_detect(all_hours[i],"^2[0-3]:[0-5][0-9]$") #5
+  }
+  if(all_hours_answers[i]==TRUE){
+    print(all_hours[i])
+  }
+  else{}
+}
+
+```
+
+Briefly, in the above code, all_hours is a set of made up and real military times with some correct and some incorrect answers, whilst all_hours_answers is a vector where the entries are TRUE or FALSE depending on whether the corresponding cell in all_hours fit the criteria of the problem or not. The #1 if statement filters out any entries starting with a 1, and #2 if statement further narrows that down to any cases between 12:01 and 12:09, since this is the only case where a 0 is not allowed in the last place, which is checked by #3. #4 checks all other military times starting with 1, and #5 checks for all military times starting with 2. In each case we use ^ to specify beginning and $ to specify end so as to pickup the 5 digits only that is standard for military time. Finally, the for loop prints which times in our vector came back true.
+
+
+## Species Names
+
+Although similar to the above case, this one is much simpler. The requirements for this one are that the first character be a captial letter, the second character a period, the third character be a space, and the fourth character onwards be only lower case letters. The regex below achieves that by allowing lower case letters to be repeated indefinitely for the species name.
+
+```{r}
+
+spec_names <- c("H. sapien", "H. Sapien", "Ho. sapien", "H sapien", "H. sapieN", "hO. sapien")
+spec_names[which(str_detect(spec_names,"^[[:upper:]].[ ][[:lower:]]+$"))]
+
+```
+
+
+## SSIDs
+
+Again, similar to the military time question, SSIDs have a unique pattern of three numeric, hyphen, two numeric, hypen, four numeric, end. The below acheives that by using {} to specify the number of repeats and ^ and $ to specify the beginning and the end of the regex respectively.
+
+```{r}
+
+ssids <- c("123-45-6789","1234-56-7890", "12-34-5678", "123-456-7890", "123-4-5678", "123-45-67890", "123-45-678")
+ssids[which(str_detect(ssids,"^[[:digit:]]{3}-[[:digit:]]{2}-[[:digit:]]{4}$"))]

```

Comment:
Good job here but just for future reference this could have been simplified with:

regex1="(1[2-9]|2[0-3]):[0-5][0-9]"

testset=c('09:30','01:55','cat','15:45','22:15')

str_detect(testset,regex1)

regex2="[A-Z]\\. [a-z]+"

testset=c('dog','H. sapien','blue jay','M. salmoides','L. lota')

str_detect(testset,regex2)

regex3="[0-9]{3}-[0-9]{2}-[0-9]{4}"

testset=c('1112223333','111-22-3333','123456789','123-45-6789')

str_detect(testset,regex3)